### PR TITLE
[REVIEW] Bug fixes and tests for subset_trajectory_id()

### PR DESCRIPTION
**Pull request by harrism**
This PR fixes multiple issues in `trajectory_subset_id`:

 - Uses `cudf::timestamp` rather than `cuspatial::its_timestamp`, since the latter can't be stored in a `gdf_column`
 - Uses signed integers including `gdf_size_type` where appropriate
 - Tests are simplified, more comprehensive, and cleaner; and they actually assert correctness of the results.
 - Adds Cython for this function.

### CODE REVIEW

File: cpp/src/trajectory/subset_trajectories.cu [hunk 1]
```diff
@@ -45,72 +46,67 @@ struct subset_functor {
...
-        thrust::sort(exec_policy, temp_id.begin(), temp_id.end());            
-        thrust::device_vector<bool> hit_vec(num_rec);
-        thrust::binary_search(exec_policy, temp_id.cbegin(), temp_id.cend(),
-                              in_id_ptr, in_id_ptr + num_rec, hit_vec.begin());
-
-
-        uint32_t num_hit = thrust::count(exec_policy, hit_vec.begin(),
-                                            hit_vec.end(), true);
+        if (num_id > 0 && id.data != nullptr && num_rec > 0) {
+            int32_t* in_id_ptr = static_cast<int32_t*>(in_id.data);
```
Comment (jrhemstad) [inline cpp/src/trajectory/subset_trajectories.cu]: The `CUDF_EXPECTS` for `in_id.dtype` allows it to be int8/16/32/64, but this is always casting to `int32_t`.

Something should change here, and I'm guessing it's making the `CUDF_EXPECTS` more restrictive.

File: cpp/src/trajectory/subset_trajectories.cu [hunk 2]
```diff
@@ -120,37 +116,41 @@ struct subset_functor {
...
+                 "Null input data");
     CUDF_EXPECTS(in_x.size == in_y.size && in_x.size == in_id.size &&
-                 in_x.size == in_timestamp.size ,
-                 "x/y/in_id/timestamp must have equal size");
-    
-    // future versions might allow x/y/in_id/timestamp to have null_count > 0,
-    // which might be useful for taking query results as inputs 
+                 in_x.size == in_timestamp.size,
+                 "Data size mismatch");
+    CUDF_EXPECTS(in_id.dtype == GDF_INT8 || in_id.dtype == GDF_INT16 ||
```
Comment (jrhemstad) [inline cpp/src/trajectory/subset_trajectories.cu]: In `subset_functor`, `in_id` is always being casted to `int32_t`, which isn't valid if the type is anything other than `GDF_INT32`.

File: cpp/src/trajectory/subset_trajectories.cu [hunk 3]
```diff
@@ -45,72 +46,67 @@ struct subset_functor {
...
+            if (num_hit > 0) {
+                out_x = cudf::allocate_like(in_x, num_hit);
+                out_y = cudf::allocate_like(in_y, num_hit);
+                out_id = cudf::allocate_like(in_id, num_hit);
+                out_timestamp = cudf::allocate_like(in_timestamp, num_hit);
+
+                auto in_itr = thrust::make_zip_iterator(thrust::make_tuple(
+                    static_cast<T*>(in_x.data), static_cast<T*>(in_y.data),
+                    static_cast<int32_t*>(in_id.data),
+                    static_cast<cuspatial::its_timestamp*>(in_timestamp.data)));
```
Comment (jrhemstad) [inline cpp/src/trajectory/subset_trajectories.cu]: Is this supposed to be `cudf::timestamp`?

File: cpp/src/trajectory/subset_trajectories.cu [hunk 4]
```diff
@@ -45,72 +46,67 @@ struct subset_functor {
...
+                auto out_itr = thrust::make_zip_iterator(thrust::make_tuple(
+                    static_cast<T*>(out_x.data), static_cast<T*>(out_y.data),
+                    static_cast<int32_t*>(out_id.data),
+                    static_cast<cuspatial::its_timestamp*>(out_timestamp.data)));
```
Comment (jrhemstad) [inline cpp/src/trajectory/subset_trajectories.cu]: Likewise, is this supposed to be `cudf::timestamp`?

File: cpp/src/trajectory/subset_trajectories.cu [hunk 5]
```diff
@@ -45,72 +46,67 @@ struct subset_functor {
...
+
+                gdf_size_type num_keep = thrust::copy_if(exec_policy, in_itr,
```
Comment (jrhemstad) [inline cpp/src/trajectory/subset_trajectories.cu]: I'd break this in to two lines. It's potentially confusing to have the subtraction of `out_itr` in the same line. 
```suggestion
               auto end  = thrust::copy_if(exec_policy, in_itr,
               gdf_size_type num_keep = end - out_itr;
```

File: cpp/src/trajectory/subset_trajectories.cu (hunk 3 above)
Comment (harrism) [reply cpp/src/trajectory/subset_trajectories.cu]: Fixed.

File: cpp/src/trajectory/subset_trajectories.cu (hunk 4 above)
Comment (harrism) [reply cpp/src/trajectory/subset_trajectories.cu]: Fixed.

File: cpp/src/trajectory/subset_trajectories.cu (hunk 5 above)
Comment (harrism) [reply cpp/src/trajectory/subset_trajectories.cu]: Good idea. Fixed.

File: cpp/src/trajectory/subset_trajectories.cu (hunk 2 above)
Comment (harrism) [reply cpp/src/trajectory/subset_trajectories.cu]: Fixed.

File: cpp/src/trajectory/subset_trajectories.cu (hunk 1 above)
Comment (harrism) [reply cpp/src/trajectory/subset_trajectories.cu]: Fixed.

**Pull request merged**